### DISCUSSION TITLE
Redirection on server actions

### DISCUSSION THREAD

**Discussion by Lzadhito** [vercel/next.js#60864] [Ideas]
### Goals

1. An option to do redirect replace on server actions
2. Function to navigate back to previous page on server actions (optional)

### Background

I've been wondering how can I fully utilize server actions for mutations. I encountered a problem which I do a DELETE on a data detail page (ex: `blog/will_deleted_id`), after that I want to either replace my browser history into another page or navigate back to previous page (ex: `blog`). With current `redirect` function on server action, [the only opt is to push](https://nextjs.org/docs/app/api-reference/functions/redirect#parameters) a page which results a navigate back into a 404 (`blog/wil_deleted_id` page already deleted).

The workaround I used for this is to use navigation on client side which not as convenience as doing post mutation functions on server

### Proposal

Enable second argument of redirect function that works on server action, or a function to navigate back on server actions

### Example
```js
"use server";
import { redirect, RedirectType } from "next/navigation";

export async function deletePost(formData: FormData) {
  const id = formData.get("id");
  await fetch(`https://delete-post.com/${id}`, { method: "DELETE" });
  redirect("/blog", RedirectType.replace);
}
```

while [The type parameter has no effect when used in Server Components](https://nextjs.org/docs/app/api-reference/functions/redirect#parameters) this example will push the browser history regardless the second param given

**Comment by ezeparziale** [comment]
Can you share an example?

**Reply by Lzadhito** [reply]
Thanks for response, I've added the example it looks like these
```js
"use server";
import { redirect, RedirectType } from "next/navigation";

export async function deletePost(formData: FormData) {
  const id = formData.get("id");
  await fetch(`https://delete-post.com/${id}`, { method: "DELETE" });
  redirect("/blog", RedirectType.replace);
}
```

while [The type parameter has no effect when used in Server Components](https://nextjs.org/docs/app/api-reference/functions/redirect#parameters) this example will push the browser history regardless the second param given.

Meanwhile I'm proposing that `redirect("/blog", RedirectType.replace)` the second param will have effect on the function to replace current browser stack

**Reply by ezeparziale** [reply]
I think that the best way is use the redirect function in the component, in this way redirect use the Replace

**Reply by Lzadhito** [reply]
That's my workaround for this case, but can you explain me why using it on component could be the best way? why not also enable it on server component?

**Reply by litehacker** [reply]
@Lzadhito Just a question - why have you classified this as `Idea` but not a `Bug`. According to their documentation it should replace, but works as push. Even if we use `"replace"` type or not. I confirm this issue on  `V 14.2.5`.

> 'replace' (default) or 'push' (default in Server Actions) 

source [redirect() Parameters](https://nextjs.org/docs/app/api-reference/functions/redirect#parameters)

**Comment by hongz1** [comment]
I am also very confused on the second parameter in server action. Although I passed second parameter with 'RedirectType.replace', server action always ignore the second parameter and it pushes so that I can go back to previous page.

As @ezeparziale mentioned, I can just use 'useRouter' and replace it in the client component, but it keeps bothering me that the usage of redirect optional parameter in server action....

**Reply by andrew310** [reply]
This is happening for me as well. I am using the `redirect` function within a server action, and it ignores the type and seems to always push. This is using next@14.2.3.

**Reply by litehacker** [reply]
> This is happening for me as well. I am using the `redirect` function within a server action, and it ignores the type and seems to always push. This is using next@14.2.3.

Let's change the issue to a "Bug"

**Reply by km-tr** [reply]
@litehacker 

I have confirmed this issue in v14.2.17.
It appears to be fixed in v15.

- https://github.com/vercel/next.js/pull/70279

**Comment by litehacker** [comment]
I believe this issue should be classified as a "Bug" rather than an "Idea" because it describes a malfunction in the existing functionality. Could we please update the issue type accordingly?

**Comment by mp3por** [comment]
Any updates on this ?

**Comment by ziyafenn** [comment]
Can confirm,`type` param also has no effect in server actions.
